TestBasicFlowsWithOriginatingIdentity go test -race detected

in https://service-catalog-jenkins.appspot.com/job/service-catalog-PR-testing2/2249/console

```
[Build & Unit/Integration Tests] === RUN   TestBasicFlowsWithOriginatingIdentity
[Build & Unit/Integration Tests] [restful] 2017/09/26 14:47:10 log.go:33: [restful/swagger] listing is available at https:///swaggerapi
[Build & Unit/Integration Tests] [restful] 2017/09/26 14:47:10 log.go:33: [restful/swagger] https:///swaggerui/ is mapped to folder /swagger-ui/
[Build & Unit/Integration Tests] E0926 14:47:12.429886    8780 controller_instance.go:1171] Failed to update status for ServiceInstance test-namespace/test-instance: Operation cannot be fulfilled on serviceinstances "test-instance": the object has been modified; please apply your changes to the latest version and try again
[Build & Unit/Integration Tests] ==================
[Build & Unit/Integration Tests] WARNING: DATA RACE
[Build & Unit/Integration Tests] Write at 0x00c420214b10 by goroutine 1034:
[Build & Unit/Integration Tests]   runtime.mapassign_faststr()
[Build & Unit/Integration Tests]       /usr/local/go/src/runtime/hashmap_fast.go:598 +0x0
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature.(*featureGate).Set()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature/feature_gate.go:136 +0x28c
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/test/integration.TestBasicFlowsWithOriginatingIdentity()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/test/integration/controller_test.go:946 +0x29f6
[Build & Unit/Integration Tests]   testing.tRunner()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:746 +0x16c
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Previous read at 0x00c420214b10 by goroutine 1145:
[Build & Unit/Integration Tests]   runtime.mapaccess2_faststr()
[Build & Unit/Integration Tests]       /usr/local/go/src/runtime/hashmap_fast.go:315 +0x0
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature.(*featureGate).Enabled()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature/feature_gate.go:182 +0x11b
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).reconcileServiceInstanceCredential()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller_binding.go:275 +0x4ad9
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).reconcileServiceInstanceCredentialKey()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller_binding.go:66 +0x2d8
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).(github.com/kubernetes-incubator/service-catalog/pkg/controller.reconcileServiceInstanceCredentialKey)-fm()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:178 +0x55
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.worker.func1.1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:209 +0xf7
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.worker.func1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:226 +0xb2
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0x6f
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98 +0xcd
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x5a
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Goroutine 1034 (running) created at:
[Build & Unit/Integration Tests]   testing.(*T).Run()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:789 +0x568
[Build & Unit/Integration Tests]   testing.runTests.func1()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:1004 +0xa7
[Build & Unit/Integration Tests]   testing.tRunner()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:746 +0x16c
[Build & Unit/Integration Tests]   testing.runTests()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:1002 +0x521
[Build & Unit/Integration Tests]   testing.(*M).Run()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:921 +0x206
[Build & Unit/Integration Tests]   main.main()
[Build & Unit/Integration Tests]       github.com/kubernetes-incubator/service-catalog/test/integration/_test/_testmain.go:70 +0x1d3
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Goroutine 1145 (running) created at:
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).Run()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:178 +0x6d6
[Build & Unit/Integration Tests] ==================
[Build & Unit/Integration Tests] ==================
[Build & Unit/Integration Tests] WARNING: DATA RACE
[Build & Unit/Integration Tests] Write at 0x00c424c9ac68 by goroutine 1034:
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature.(*featureGate).Set()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature/feature_gate.go:136 +0x2a5
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/test/integration.TestBasicFlowsWithOriginatingIdentity()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/test/integration/controller_test.go:946 +0x29f6
[Build & Unit/Integration Tests]   testing.tRunner()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:746 +0x16c
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Previous read at 0x00c424c9ac68 by goroutine 1145:
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature.(*featureGate).Enabled()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature/feature_gate.go:182 +0x13a
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).reconcileServiceInstanceCredential()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller_binding.go:275 +0x4ad9
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).reconcileServiceInstanceCredentialKey()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller_binding.go:66 +0x2d8
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).(github.com/kubernetes-incubator/service-catalog/pkg/controller.reconcileServiceInstanceCredentialKey)-fm()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:178 +0x55
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.worker.func1.1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:209 +0xf7
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.worker.func1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:226 +0xb2
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0x6f
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98 +0xcd
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x5a
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Goroutine 1034 (running) created at:
[Build & Unit/Integration Tests]   testing.(*T).Run()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:789 +0x568
[Build & Unit/Integration Tests]   testing.runTests.func1()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:1004 +0xa7
[Build & Unit/Integration Tests]   testing.tRunner()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:746 +0x16c
[Build & Unit/Integration Tests]   testing.runTests()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:1002 +0x521
[Build & Unit/Integration Tests]   testing.(*M).Run()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:921 +0x206
[Build & Unit/Integration Tests]   main.main()
[Build & Unit/Integration Tests]       github.com/kubernetes-incubator/service-catalog/test/integration/_test/_testmain.go:70 +0x1d3
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Goroutine 1145 (running) created at:
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).Run()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:178 +0x6d6
[Build & Unit/Integration Tests] ==================
[Build & Unit/Integration Tests] --- FAIL: TestBasicFlowsWithOriginatingIdentity (4.95s)
[Build & Unit/Integration Tests] 	framework.go:83: Starting server on port: 14945
[Build & Unit/Integration Tests] 	controller_test.go:1001: controller start
[Build & Unit/Integration Tests] 	controller_test.go:1009: informers start
[Build & Unit/Integration Tests] 	framework.go:79: Shutting down server on port: 14945
[Build & Unit/Integration Tests] 	testing.go:699: race detected during execution of test
[Build & Unit/Integration Tests] FAIL
```

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -170,13 +170,22 @@ func (c *controller) Run(workers int, stopCh <-chan struct{}) {
 
 	glog.Info("Starting service-catalog controller")
 
+	totalNumberOfWorkers := 0
+	done := make(chan struct{})
+
 	for i := 0; i < workers; i++ {
-		go wait.Until(worker(c.brokerQueue, "ServiceBroker", maxRetries, true, c.reconcileServiceBrokerKey), time.Second, stopCh)
-		go wait.Until(worker(c.serviceClassQueue, "ServiceClass", maxRetries, true, c.reconcileServiceClassKey), time.Second, stopCh)
-		go wait.Until(worker(c.servicePlanQueue, "ServicePlan", maxRetries, true, c.reconcileServicePlanKey), time.Second, stopCh)
-		go wait.Until(worker(c.instanceQueue, "ServiceInstance", maxRetries, true, c.reconcileServiceInstanceKey), time.Second, stopCh)
-		go wait.Until(worker(c.bindingQueue, "ServiceInstanceCredential", maxRetries, true, c.reconcileServiceInstanceCredentialKey), time.Second, stopCh)
-		go wait.Until(worker(c.pollingQueue, "Poller", maxRetries, false, c.requeueServiceInstanceForPoll), time.Second, stopCh)
+		createWorker(c.brokerQueue, "ServiceBroker", maxRetries, true, c.reconcileServiceBrokerKey, stopCh, done)
+		totalNumberOfWorkers++
+		createWorker(c.serviceClassQueue, "ServiceClass", maxRetries, true, c.reconcileServiceClassKey, stopCh, done)
+		totalNumberOfWorkers++
+		createWorker(c.servicePlanQueue, "ServicePlan", maxRetries, true, c.reconcileServicePlanKey, stopCh, done)
+		totalNumberOfWorkers++
+		createWorker(c.instanceQueue, "ServiceInstance", maxRetries, true, c.reconcileServiceInstanceKey, stopCh, done)
+		totalNumberOfWorkers++
+		createWorker(c.bindingQueue, "ServiceInstanceCredential", maxRetries, true, c.reconcileServiceInstanceCredentialKey, stopCh, done)
+		totalNumberOfWorkers++
+		createWorker(c.pollingQueue, "Poller", maxRetries, false, c.requeueServiceInstanceForPoll, stopCh, done)
+		totalNumberOfWorkers++
 	}
 
 	<-stopCh
@@ -188,6 +197,20 @@ func (c *controller) Run(workers int, stopCh <-chan struct{}) {
 	c.instanceQueue.ShutDown()
 	c.bindingQueue.ShutDown()
 	c.pollingQueue.ShutDown()
+
+	for i := 0; i < totalNumberOfWorkers; i++ {
+		<-done
+	}
+}
+
+// createWorker creates and runs a worker thread that just processes items in the
+// specified queue. The worker will run until stopCh is closed. When the worker
+// stops, a token will be added to the done channel.
+func createWorker(queue workqueue.RateLimitingInterface, resourceType string, maxRetries int, forgetAfterSuccess bool, reconciler func(key string) error, stopCh <-chan struct{}, done chan<- struct{}) {
+	go func() {
+		wait.Until(worker(queue, resourceType, maxRetries, forgetAfterSuccess, reconciler), time.Second, stopCh)
+		done <- struct{}{}
+	}()
 }
 
 // worker runs a worker thread that just dequeues items, processes them, and marks them done.

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -19,6 +19,7 @@ package controller
 import (
 	"encoding/json"
 	"fmt"
+	"sync"
 	"time"
 
 	"github.com/golang/glog"
@@ -170,22 +171,15 @@ func (c *controller) Run(workers int, stopCh <-chan struct{}) {
 
 	glog.Info("Starting service-catalog controller")
 
-	totalNumberOfWorkers := 0
-	done := make(chan struct{})
+	var waitGroup sync.WaitGroup
 
 	for i := 0; i < workers; i++ {
-		createWorker(c.brokerQueue, "ServiceBroker", maxRetries, true, c.reconcileServiceBrokerKey, stopCh, done)
-		totalNumberOfWorkers++
-		createWorker(c.serviceClassQueue, "ServiceClass", maxRetries, true, c.reconcileServiceClassKey, stopCh, done)
-		totalNumberOfWorkers++
-		createWorker(c.servicePlanQueue, "ServicePlan", maxRetries, true, c.reconcileServicePlanKey, stopCh, done)
-		totalNumberOfWorkers++
-		createWorker(c.instanceQueue, "ServiceInstance", maxRetries, true, c.reconcileServiceInstanceKey, stopCh, done)
-		totalNumberOfWorkers++
-		createWorker(c.bindingQueue, "ServiceInstanceCredential", maxRetries, true, c.reconcileServiceInstanceCredentialKey, stopCh, done)
-		totalNumberOfWorkers++
-		createWorker(c.pollingQueue, "Poller", maxRetries, false, c.requeueServiceInstanceForPoll, stopCh, done)
-		totalNumberOfWorkers++
+		createWorker(c.brokerQueue, "ServiceBroker", maxRetries, true, c.reconcileServiceBrokerKey, stopCh, &waitGroup)
+		createWorker(c.serviceClassQueue, "ServiceClass", maxRetries, true, c.reconcileServiceClassKey, stopCh, &waitGroup)
+		createWorker(c.servicePlanQueue, "ServicePlan", maxRetries, true, c.reconcileServicePlanKey, stopCh, &waitGroup)
+		createWorker(c.instanceQueue, "ServiceInstance", maxRetries, true, c.reconcileServiceInstanceKey, stopCh, &waitGroup)
+		createWorker(c.bindingQueue, "ServiceInstanceCredential", maxRetries, true, c.reconcileServiceInstanceCredentialKey, stopCh, &waitGroup)
+		createWorker(c.pollingQueue, "Poller", maxRetries, false, c.requeueServiceInstanceForPoll, stopCh, &waitGroup)
 	}
 
 	<-stopCh
@@ -198,18 +192,17 @@ func (c *controller) Run(workers int, stopCh <-chan struct{}) {
 	c.bindingQueue.ShutDown()
 	c.pollingQueue.ShutDown()
 
-	for i := 0; i < totalNumberOfWorkers; i++ {
-		<-done
-	}
+	waitGroup.Wait()
 }
 
 // createWorker creates and runs a worker thread that just processes items in the
-// specified queue. The worker will run until stopCh is closed. When the worker
-// stops, a token will be added to the done channel.
-func createWorker(queue workqueue.RateLimitingInterface, resourceType string, maxRetries int, forgetAfterSuccess bool, reconciler func(key string) error, stopCh <-chan struct{}, done chan<- struct{}) {
+// specified queue. The worker will run until stopCh is closed. The worker will be
+// added to the wait group when started and marked done when finished.
+func createWorker(queue workqueue.RateLimitingInterface, resourceType string, maxRetries int, forgetAfterSuccess bool, reconciler func(key string) error, stopCh <-chan struct{}, waitGroup *sync.WaitGroup) {
+	waitGroup.Add(1)
 	go func() {
 		wait.Until(worker(queue, resourceType, maxRetries, forgetAfterSuccess, reconciler), time.Second, stopCh)
-		done <- struct{}{}
+		waitGroup.Done()
 	}()
 }
 

--- a/test/integration/controller_test.go
+++ b/test/integration/controller_test.go
@@ -86,7 +86,7 @@ func truePtr() *bool {
 //
 // ...using purely synchronous provision/deprovision.
 func TestBasicFlowsSync(t *testing.T) {
-	_, catalogClient, _, _, _, _, shutdownServer := newTestController(t, fakeosb.FakeClientConfiguration{
+	_, catalogClient, _, _, _, _, shutdownServer, shutdownController := newTestController(t, fakeosb.FakeClientConfiguration{
 		CatalogReaction: &fakeosb.CatalogReaction{
 			Response: getTestCatalogResponse(),
 		},
@@ -110,6 +110,7 @@ func TestBasicFlowsSync(t *testing.T) {
 			},
 		},
 	})
+	defer shutdownController()
 	defer shutdownServer()
 
 	client := catalogClient.ServicecatalogV1alpha1()
@@ -249,7 +250,7 @@ func TestBasicFlowsSync(t *testing.T) {
 // TestBasicFlowsAsync tests the same flows as TestBasicFlowsSync, using
 // asynchronous provision/deprovision.
 func TestBasicFlowsAsync(t *testing.T) {
-	_, catalogClient, _, _, _, _, shutdownServer := newTestController(t, fakeosb.FakeClientConfiguration{
+	_, catalogClient, _, _, _, _, shutdownServer, shutdownController := newTestController(t, fakeosb.FakeClientConfiguration{
 		CatalogReaction: &fakeosb.CatalogReaction{
 			Response: getTestCatalogResponse(),
 		},
@@ -278,6 +279,7 @@ func TestBasicFlowsAsync(t *testing.T) {
 			},
 		},
 	})
+	defer shutdownController()
 	defer shutdownServer()
 
 	client := catalogClient.ServicecatalogV1alpha1()
@@ -420,7 +422,7 @@ func TestBasicFlowsAsync(t *testing.T) {
 // TODO: additional tests for scenarios like this will be needed once we
 // implement orphan mitigation.
 func TestProvisionFailure(t *testing.T) {
-	_, catalogClient, _, _, _, _, shutdownServer := newTestController(t, fakeosb.FakeClientConfiguration{
+	_, catalogClient, _, _, _, _, shutdownServer, shutdownController := newTestController(t, fakeosb.FakeClientConfiguration{
 		CatalogReaction: &fakeosb.CatalogReaction{
 			Response: getTestCatalogResponse(),
 		},
@@ -435,6 +437,7 @@ func TestProvisionFailure(t *testing.T) {
 		// return an unexpected call error message if deprovision is called on
 		// the broker.
 	})
+	defer shutdownController()
 	defer shutdownServer()
 
 	client := catalogClient.ServicecatalogV1alpha1()
@@ -536,7 +539,7 @@ func TestProvisionFailure(t *testing.T) {
 // TestBindingFailure tests that a binding gets a failure condition when the
 // broker returns a failure response for a bind operation.
 func TestBindingFailure(t *testing.T) {
-	_, fakeCatalogClient, _, _, _, _, shutdownServer := newTestController(t, fakeosb.FakeClientConfiguration{
+	_, fakeCatalogClient, _, _, _, _, shutdownServer, shutdownController := newTestController(t, fakeosb.FakeClientConfiguration{
 		CatalogReaction: &fakeosb.CatalogReaction{
 			Response: getTestCatalogResponse(),
 		},
@@ -559,6 +562,7 @@ func TestBindingFailure(t *testing.T) {
 			},
 		},
 	})
+	defer shutdownController()
 	defer shutdownServer()
 
 	client := fakeCatalogClient.ServicecatalogV1alpha1()
@@ -702,7 +706,7 @@ func TestBasicFlowsWithOriginatingIdentity(t *testing.T) {
 	utilfeature.DefaultFeatureGate.Set(fmt.Sprintf("%v=true", scfeatures.OriginatingIdentity))
 	defer utilfeature.DefaultFeatureGate.Set(fmt.Sprintf("%v=false", scfeatures.OriginatingIdentity))
 
-	_, catalogClient, catalogClientConfig, _, _, _, shutdownServer := newTestController(t, fakeosb.FakeClientConfiguration{
+	_, catalogClient, catalogClientConfig, _, _, _, shutdownServer, shutdownController := newTestController(t, fakeosb.FakeClientConfiguration{
 		CatalogReaction: &fakeosb.CatalogReaction{
 			Response: &osb.CatalogResponse{
 				Services: []osb.Service{
@@ -743,6 +747,7 @@ func TestBasicFlowsWithOriginatingIdentity(t *testing.T) {
 			},
 		},
 	})
+	defer shutdownController()
 	defer shutdownServer()
 
 	client := catalogClient.ServicecatalogV1alpha1()
@@ -963,6 +968,7 @@ func newTestController(t *testing.T, config fakeosb.FakeClientConfiguration) (
 	*fakeosb.FakeClient,
 	controller.Controller,
 	informers.Interface,
+	func(),
 	func()) {
 
 	// create a fake kube client
@@ -1004,10 +1010,20 @@ func newTestController(t *testing.T, config fakeosb.FakeClientConfiguration) (
 	}
 
 	stopCh := make(chan struct{})
-	go testController.Run(1, stopCh)
+	controllerStopped := make(chan struct{})
+	go func() {
+		testController.Run(1, stopCh)
+		controllerStopped <- struct{}{}
+	}()
 	informerFactory.Start(stopCh)
 	t.Log("informers start")
-	return fakeKubeClient, catalogClient, catalogClientConfig, fakeOSBClient, testController, serviceCatalogSharedInformers, shutdownServer
+
+	shutdownController := func() {
+		close(stopCh)
+		<-controllerStopped
+	}
+
+	return fakeKubeClient, catalogClient, catalogClientConfig, fakeOSBClient, testController, serviceCatalogSharedInformers, shutdownServer, shutdownController
 }
 
 func changeUsernameForCatalogClient(catalogClient clientset.Interface, catalogClientConfig *restclient.Config, username string) (clientset.Interface, error) {

